gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagementTaskStreamImpl.java

```diff
@@ -0,0 +1,159 @@
+    }
+    this.isActive = active;
+    try {
+      if (this.isActive) {
+        log.info("Activating DagManagementTaskStreamImpl.");
+        //Initializing state store for persisting Dags.
+        this.dagManagementStateStore.start();
+        dagManagerMetrics.activate();
+      } else { //Mark the DagManager inactive.
+        log.info("Inactivating the DagManagementTaskStreamImpl. Shutting down all DagManager threads");
+        dagManagerMetrics.cleanup();
+      }
+    } catch (IOException e) {
+        log.error("Exception encountered when activating the DagManagementTaskStreamImpl", e);
+        throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public synchronized void addDagAction(DagActionStore.DagAction dagAction) throws IOException {
+    // TODO: Used to track missing dag issue, remove later as needed
+    log.info("Add dagAction{}", dagAction);
+    if (!isActive) {
+      log.warn("Skipping add dagAction because this instance of DagManagementTaskStreamImpl is not active for dag: {}",
+          dagAction);
+      return;
+    }
+
+    DagManager.DagId dagId = DagManagerUtils.generateDagId(dagAction.getFlowGroup(), dagAction.getFlowName(), dagAction.getFlowExecutionId());
+    String dagIdString = dagId.toString();
+    if (this.dagManagementStateStore.containsDag(dagId)) {
+      log.warn("Already tracking a dag with dagId {}, skipping.", dagIdString);
+      return;
+    }
+
+    // After persisting the dag, its status will be tracked by active dagManagers so the action should be deleted
+    // to avoid duplicate executions upon leadership change
+    if (this.dagActionStore.isPresent()) {
+      this.dagActionStore.get().deleteDagAction(dagAction);
+    }
+
+    if (!this.dagActionQueue.offer(dagAction)) {
+      throw new RuntimeException("Could not add dag action " + dagAction + " to the queue");
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !this.dagActionQueue.isEmpty();
+  }
+
+  @Override
+  public DagTask<DagProc> next() {
+    try {
+      DagActionStore.DagAction dagAction = this.dagActionQueue.take();  //`take` blocks till element is not available
+      // todo reconsider the use of MultiActiveLeaseArbiter
+      //MultiActiveLeaseArbiter.LeaseAttemptStatus leaseAttemptStatus = new MultiActiveLeaseArbiter.LeaseObtainedStatus(dagAction);
+      // todo - uncomment after flow trigger handler provides such an api
+      //Properties jobProps = getJobProperties(dagAction);
+      //flowTriggerHandler.getLeaseOnDagAction(jobProps, dagAction, System.currentTimeMillis());
+      //if (leaseAttemptStatus instanceof MultiActiveLeaseArbiter.LeaseObtainedStatus) {
+      // can it return null? is this iterator allowed to return null?
+      return createDagTask(dagAction, new MultiActiveLeaseArbiter.LeaseObtainedStatus(dagAction, System.currentTimeMillis()));
+      //}
+    } catch (Throwable t) {
```

Comment:
catching `Throwable` is pretty extreme, given it would swallow OOM and other low-level `Error`s.  what specifically are the kinds of exceptions you're on the lookout for?